File: src/FASTCAMPUS/ALGO/D36/P20364.js
const fs = require('fs');
class Scanner {
  constructor(fileName){
    this.read = fs.readFileSync(fileName).toString().trim().split('\n');
    this.stringToken = [];
  }
  next(){
    if(this.stringToken.length === 0){
      if(this.read.length > 0){
        this.stringToken = this.read.shift().split(" ");
      } else {
        return null;
      }
    } 
    return this.stringToken.shift();
  }

  nextNumber(){
    const next = this.next();
    if(next){
      return Number(next);
    } else {
      return null;
    }
    
  }

  nextLine(){
    if(this.read.length > 0){
      return this.read.shift();
    } else {
      return null;
    }
    
  }
}

const scanner = new Scanner('stdin.txt');
let N, Q, duck;
let tree;
const input = function (){
  N = scanner.nextNumber();
  Q = scanner.nextNumber();
  tree = Array.from({length: N + 1}, () => false);
  duck = scanner.read.map(Number)
}

const pro = function (){
  console.log(tree);
  console.log(duck);
  duck.forEach(d => { 
    let res = 0;
    let c = d;
    while(d > 0){
      if(tree[d]) res = d;
      d /= 2;
    }
    tree[c] = true;
    console.log(res);
  })
}

const solve = function(){
  input();
  pro();
}

solve();

File: src/FASTCAMPUS/ALGO/D36/P20364.java
package FASTCAMPUS.ALGO.D36;
import java.io.*;
import java.util.*;

public class P20364 {
    static int N, Q;
    static int [] duck;
    static boolean[] tree;
    static StringBuilder sb;
    static void input() {
        FastReader scan = new FastReader();
        N = scan.nextInt();
        Q = scan.nextInt();
        tree = new boolean [N + 1];
        duck = new int [Q];
        for(int i = 0; i < Q; i ++){
            duck[i] = scan.nextInt();
        }
        sb = new StringBuilder();
    }
    public static void pro(){
        for (int i : duck) { 
            int res = 0;
            int c = i;
            while(i > 0){
             if(tree[i]) res = i;
                i /= 2;
            }
            tree[c] = true;
            sb.append(res);
            sb.append("\n");
        }
    }
    public static void main(String[] args) {
        input();
        pro();
        System.out.println(sb);
    }
    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        public FastReader(String s) throws FileNotFoundException {
            br = new BufferedReader(new FileReader(new File(s)));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}